src/users.rs

```diff
@@ -0,0 +1,243 @@
+                    .prepare("SELECT id, name, salt, hash, active FROM users WHERE name = :name LIMIT 1;")
+                    .expect("");
+                // Run the Query Passing in the user name
+                let mut rows = stmt
+                    .query_named(named_params! { ":name": self.name })
+                    .expect("");
+                // Iterate over results (Should only ever return one)
+                while let Some(row) = rows.next().expect("") {
+                    // Set Salt from DB
+                    self.salt = row.get(2)?;
+                    // Set Hash From DB
+                    self.hash = row.get(3)?;
+                    // Set Active From DB
+                    let active: i32 = row.get(4)?;
+                    self.active = active != 0;
+                }
+                // Decode Hash into bytes
+                let hash = HEXUPPER
+                    .decode(self.hash.as_bytes())
+                    .expect("Error Decoding Hash");
+                // Decode Salt into bytes
+                let salt = HEXUPPER
+                    .decode(self.salt.as_bytes())
+                    .expect("Error Decoding Salt");
+                // Verify Password against hash using salt
+                match verify_password(password, salt, hash) {
+                    Ok(()) => {
+                        return Ok(());
+                    }
+                    Err(_) => {
+                        return Err(Box::new(UserError("Invalid User".into())));
+                    }
+                }
+            }
+            _ => return Err(Box::new(UserError("DB Connection Error".into()))),
+        };
+    }
+
+    // TODO: Implement update feature
+    // pub fn update(&mut self) {}
+
+    // TODO: Implement Delete Feature
+    // pub fn delete(&mut self) {}
+}
+
+#[cfg(test)]
+mod tests {
+    use super::User;
+    use super::{get_hash_from_password, get_salt, verify_password};
+    use ring::digest;
+    use std::fs;
+
+    #[test]
+    fn test_get_salt() {
+        let salt = get_salt();
+        assert!(
+            salt.len() == digest::SHA512_OUTPUT_LEN,
+            "Expected Salt to be of length {} but got {}",
+            digest::SHA512_OUTPUT_LEN,
+            salt.len()
+        )
+    }
+
+    #[test]
+    fn test_hashing_password() {
+        let salt = get_salt();
+        let hash = get_hash_from_password(String::from("Password1"), salt.clone());
+        let verify_correct_password =
+            verify_password(String::from("Password1"), salt.clone(), hash.clone());
+        let verify_incorrect_password = verify_password(String::from("Password2"), salt, hash);
+        assert!(
+            verify_correct_password.is_ok(),
+            "Expected password to get verified"
+        );
+        assert!(
+            !verify_incorrect_password.is_ok(),
+            "Expected password to fail verification"
+        )
+    }
+
+    #[test]
+    fn test_create_and_get() {
```

Comment:
Can we add some tests for failing cases? e.g. trying to get a user that doesn't exist?

Reply:
Yes sir :)